Ubuntu/xenial

blackboxsw opened the pull request:
Backport  cc_ubuntu_pro module to xenial to ensure all ESM-supported series allow using the `ubuntu_pro:` key in cloud-config user-data to align with product naming.

significant changes dealt with during this backport to Xenial:
- config schema is defined in cc_ubuntu_pro module as 'schema' var instead of `meta`
- no centralized cloudinit/config/schemas/schema-cloud-config-v1.json
 - no f-string support in xenial's python
 - unittests don't live under tests/unittests but under cloudinit/config/tests/test_ubuntu_pro
 - no pytest.param
 - inline type hints `varname: dict = {}` are not valid syntax
 - no mocker fixture in pytest
 - no jsonschema installed for unittests either 
   -  json schema validation is inert in xenial because python3-jsonschema is not a strict package Depends: so cloud-init skips any schema validation.
 - Expect no warnings if schema keys are invalid, they are just ignored in this version of cloud-init

## do not merge. Merge will be performed on the cmdline


## Proposed Commit Message
```
    feat(pro): deprecate ubuntu_advantage key for ubuntu_pro
    
    Backport deprecatation of ubuntu_advantage_key in favor of ubuntu_pro
    to standardize product branding for Ubuntu Pro.
    
    This allows for common behavior, documentation and tooling on all
    
    Backport deprecatation of ubuntu_advantage_key in favor of ubuntu_pro
    to standardize product branding for Ubuntu Pro.
    
    This allows for common behavior, documentation and tooling on all
    supported and maintained releases including those Ubuntu releases
    under active Expanded Security Maintenance(ESM).
    
    LP: #2067660
```

## Additional Context
SRU process bug
[LP: #2067660](https://bugs.launchpad.net/ubuntu/+source/cloud-init/+bug/2067660)

## Test Steps
Validate quilt patches, and whether package can build
```
quilt push -a; tox -e py3; quilt pop -a
build-package
sbulid --dist=xenial --arch=amd64 --arch-all ../out/cloud-init_21.1-19-gbad84ad4-0ubuntu1~16.04.5.dsc
```
- SRU behavior verification:

1. Assert ubuntu_pro key is ignored on released xenial cloud-init
2. Assert ubuntu_pro key is honored on -proposed cloud-init. (pro detach/cloud-init clean)
3. Assert ubuntu-advantage key is honored on -proposed cloud-init (pro detach/cloud-init clean)
4. Assert ubuntu-advantage key debug logs about deprecation in /var/log/cloud-init.og

```
#!/bin/bash
set -ex
RELEASE=$1
VM_NAME=backport-$1-ubuntu-pro
lxc init ubuntu-daily:$RELEASE --vm $VM_NAME
lxc config device add $VM_NAME config disk source=cloud-init:config

if [ $RELEASE = "xenial" ]; then
WAIT_CMD="cloud-init status --wait --long"
lxc config set $VM_NAME cloud-init.user-data - << EOF
#cloud-config
packages:
  - ubuntu-advantage-tools
  - linux-image-virtual-hwe-16.04  # 16.04 GA kernel as a problem with vsock
runcmd:
  - mount -t 9p config /mnt
  - cd /mnt
  - ./install.sh
  - cd /
  - umount /mnt
  - systemctl start lxd-agent  # XXX: causes a reboot
EOF
else
WAIT_CMD="cloud-init status --wait --format=yaml"
lxc config set $VM_NAME cloud-init.user-data - << EOF
#cloud-config
packages:
  - ubuntu-advantage-tools
runcmd:
  - mount -t 9p config /mnt
  - cd /mnt
  - ./install.sh
  - cd /
  - umount /mnt
  - systemctl start lxd-agent  # XXX: causes a reboot
EOF
fi

lxc start --console $VM_NAME  || true # hit enter on prompt
sleep 10

echo "-- confirm cloud-init boot complete"
lxc exec $VM_NAME -- cloud-init status --wait --format=yaml

echo "-- setup ubuntu_pro attach user-data"
cat > ubuntu_pro.yaml << EOF
#cloud-config
ubuntu_pro:
  token: $UA_TOKEN
EOF

echo "-- setup ubuntu-advantage attach user-data"
cat > ubuntu_adv.yaml << EOF
#cloud-config
ubuntu-advantage:
  token: $UA_TOKEN
EOF

echo "-- setup ubuntu-advantage commands user-data"
cat > ubuntu_adv_cmd.yaml << EOF
#cloud-config
ubuntu-advantage:
  commands:
     00: [attach, $UA_TOKEN]
EOF

lxc file push ubuntu_pro.yaml $VM_NAME/var/lib/cloud/seed/nocloud-net/user-data
lxc config set $VM_NAME cloud-init.user-data="$(cat ubuntu_pro.yaml)"
lxc exec $VM_NAME -- cloud-init clean --logs --reboot || true
sleep 10

echo "-- confirm orig cloud-init ignores ubuntu_pro"
lxc exec $VM_NAME -- $WAIT_CMD
lxc exec $VM_NAME -- pro status

echo "-- Proposed cloud-init honors ubuntu_pro key"
lxc exec $VM_NAME -- add-apt-repository ppa:chad.smith/esm-backport-ubuntu-pro-config2 -y
lxc exec $VM_NAME -- apt-get install cloud-init -y
lxc exec $VM_NAME -- $WAIT_CMD
lxc exec $VM_NAME -- cloud-init clean --logs --reboot || true
sleep 10
lxc exec $VM_NAME -- $WAIT_CMD
lxc exec $VM_NAME -- pro status | grep enabled


echo "-- Proposed cloud-init still honors ubuntu-advantage key"
lxc file push ubuntu_adv.yaml $VM_NAME/var/lib/cloud/seed/nocloud-net/user-data
lxc exec $VM_NAME -- pro detach
lxc exec $VM_NAME -- pro status
lxc exec $VM_NAME -- rm -rf /var/lib/ubuntu-advantage
sed -i 's/ubuntu_pro/ubuntu-advantage/' ubuntu_pro.yaml
lxc config set $VM_NAME cloud-init.user-data="$(cat ubuntu_adv.yaml)"

lxc exec $VM_NAME -- cloud-init clean --logs --reboot || true
sleep 10
lxc exec $VM_NAME -- $WAIT_CMD
lxc exec $VM_NAME -- pro status
echo "-- Proposed cloud-init logs debug deprecation messages"
lxc exec $VM_NAME -- grep ubuntu-adv /var/log/cloud-init.log

echo "-- Proposed cloud-init still honors ubuntu-advantage:commands key"
lxc file push ubuntu_adv_cmd.yaml $VM_NAME/var/lib/cloud/seed/nocloud-net/user-data
lxc exec $VM_NAME -- pro detach
lxc exec $VM_NAME -- pro status
lxc exec $VM_NAME -- rm -rf /var/lib/ubuntu-advantage
sed -i 's/ubuntu_pro/ubuntu-advantage/' ubuntu_pro.yaml
lxc config set $VM_NAME cloud-init.user-data="$(cat ubuntu_adv_cmd.yaml)"

lxc exec $VM_NAME -- cloud-init clean --logs --reboot || true
sleep 10
lxc exec $VM_NAME -- $WAIT_CMD
lxc exec $VM_NAME -- pro status
echo "-- Proposed cloud-init logs debug deprecation messages"
lxc exec $VM_NAME -- grep ubuntu-adv /var/log/cloud-init.log


```
## Merge type

- [ ] Squash merge using "Proposed Commit Message"
- [x] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

TheRealFalcon commented on debian/patches/backport-ubuntu-pro-rename.patch:
```diff
@@ -0,0 +1,2522 @@
++        LOG.error(msg)
++        raise RuntimeError(msg)
++    validate_cloudconfig_schema(cfg, schema)
++    if "commands" in pro_section:
```
Was this behavior already broken in Xenial? If not, why are we breaking it here?

blackboxsw replied:
Great point. pulling this element in wholesales doesn't make sense to introduce this error. We should continue to support "commands" in xenial.

TheRealFalcon reviewed (changes requested):
I left a few comments inline.

blackboxsw commented on debian/patches/backport-ubuntu-pro-rename.patch:
```diff
@@ -0,0 +1,2522 @@
+-    'properties': {
+-        'ubuntu-advantage': {
+-            'type': 'object',
+-            'properties': {
```
Good thinking, I had mistakenly looked over the xenial branch before applying quilt patches and didn't recall that we carried the d/patches/ubuntu-advantage-revert-tip.patch.  I have since dropped that patch from our series and just applied the direct d/patches/backport-ubuntu-pro-config.patch which handles file renames and module renames and cc_ubunutu_pro.schema additions. I"ll update the test verification  in the PR to also exercise the ubuntu-advantage:commands testing.

blackboxsw commented:
@TheRealFalcon, for next week, I believe I had addressed your commands, dropped our original debian/patches/ubuntu-advantage-revert-tip.patch and included only the new backport-ubuntu-pro-config.patch since there is no sense applying changes in an earlier patch only to alter those changes once again in the backport and rename of cc_ubuntu_advantage.py  to cc_ubuntu_pro.py

TheRealFalcon commented on debian/patches/backport-ubuntu-pro-config.patch:
```diff
@@ -0,0 +1,2714 @@
++        }
++      }
++    },
++    "config": {
```
I don't understand why these are here. Does Pro on Xenial do anything with these? Same question with the `KNOWN_PRO_CONFIG_PROPS`.

blackboxsw replied:
@TheRealFalcon the Pro client on xenial does support`pro config (set|unset)` as latest Pro client version 27.5 is SRU'd t xenial already so latest pro client cli is already supported there. With this backport PR for cloud-init, Pro customers will be able to now launch cloud-config containing `ubuntu_pro: {config: {k:v, ....}}` and know that [cloud-init can speak `pro config set k=v`](https://github.com/canonical/cloud-init/pull/5600/files#diff-9a65f7ac405261b233a2462001db0bc7d4a06ef84a04edd2ddf1ec2515aa90ecR530) during initial boot.  Original Xenial didn't support this feature, but it it is additive per this backport and a non-breaking change.


Per KNOWN_PRO_CONFIG_PROPS, it is a subset of known supported `pro config` properties on all series and can grow independently as new config options are added to pro client and released everywhere.
 On xenial, pro config show for pro --version 27.5 shows that our subset in cloud-init is well within supported config key names.
```
root@test-x-vm:~# pro config show
http_proxy              None
https_proxy             None
apt_http_proxy          None
apt_https_proxy         None
ua_apt_http_proxy       None
ua_apt_https_proxy      None
global_apt_http_proxy   None
global_apt_https_proxy  None
update_messaging_timer  21600
metering_timer          14400
apt_news                True
apt_news_url            https://motd.ubuntu.com/aptnews.json
```

We can extend those KNOWN_PRO_CONFIG_PROPS keys in tip of main in the future if we want strict schema validation and concise errors in cloud-init, but it is not necessary as our JSON schema allows for extraProperties: True to remain flexible with unknown pro config keys.

TheRealFalcon reviewed (commented):
Nice updates. One question left.

TheRealFalcon reviewed (approved):
Thanks for the explanation!
